pkg/receive/writer.go

```diff
@@ -81,6 +82,14 @@ func (r *Writer) Write(ctx context.Context, tenantID string, wreq *prompb.WriteR
 			lset = labelpb.ZLabelsToPromLabels(t.Labels)
 		}
 
+		// If the ref is 0, it indicates that inserting the samples will create a new time series.
+		// We do the seriesLimit check before actually creating the series in head because even after
+		// a rollback, the series will stay in head, defeating the whole purpose of series limit.
+		if ref == 0 && seriesLimit > 0 && s.NumSeries() >= seriesLimit {
+			_ = app.Rollback()
```

Comment:
What are we trying to rollback in this case? We only check whether the series is new in TSDB head and we don't append anything. Do we still need this line?

Reply:
Because we don't want to do partial writes, so I am explicitly rolling back the previous appends. I think due to the early return `app.Commit()` won't be called, so no partial writes anyways, but still calling `app.Rollback()` is looks like a nice idea to me as it will close the `Appender`.